File: pset1/credit.c
// Irvan Tristian
// 19 January 2019
// Problem Set 1 - Check whether CC Number is AMEX, MASTERCARD, VISA, or INVALID
#include <stdio.h>
#include <cs50.h>

int main(void)
{
    // Define local variable within main
    long number;
    
    // Get CC Number input from the user
    do
    {
        number = get_long("Number: ");
    }
    while (number <= 0);
    
    // Define variable for Finding out length of Card Number
    long n = number;
    int len_number = 0;
    
    // Find out length of Card Number
    while (n != 0)
    {
        n /= 10;
        ++len_number;
    }
    
    // Define variable for calculating
    int count = len_number / 2;
    long num_scale = 10;
    int sum = 0;
    int digit0;
    int digit1;
    
    // Calculate from second-to-last digit
    for (int i = 0; i < count; i++)
    {
        long digit = (number / num_scale) % 10;
        int each_sum = digit * 2;
        if (each_sum >= 10)
        {
            each_sum = (each_sum % 10) + 1;
        }
        sum += each_sum;
        if (i == count - 1)
        {
            digit0 = digit;
        }
        num_scale *= 100;
    }
    
    // Redefine few variable for another calculating
    num_scale = 1;
    count = len_number - (len_number / 2);
    
    // Calculate with the sum of the other digits 
    for (int i = 0; i < count; i++)
    {
        long digit = (number / num_scale) % 10;
        sum += digit;
        if (i == count - 1)
        {
            digit1 = digit;
        }
        num_scale *= 100;
    }
    
    // Define variable for validating
    int check = sum % 10;
    
    // Validate last digit in the sum
    if (check == 0)
    {
        switch (len_number)
        {
            case 15:
                if (digit1 == 3)
                {
                    if (digit0 == 4 || digit0 == 7)
                    {
                        printf("AMEX\n");
                    }
                    else
                    {
                        printf("INVALID\n");
                    }
                }
                else
                {
                    printf("INVALID\n");
                }
                break;
                
            case 13:
                if (digit1 == 4)
                {
                    printf("VISA\n");
                }
                else
                {
                    printf("INVALID\n");
                }
                break;
                
            case 16:
                if (digit0 == 4)
                {
                    printf("VISA\n");
                }
                else if (digit0 == 5)
                {
                    if (digit1 >= 1 && digit1 <= 5)
                    {
                        printf("MASTERCARD\n");
                    }
                    else
                    {
                        printf("INVALID\n");
                    }
                }
                else
                {
                    printf("INVALID\n");
                }
                break;
            default:
                printf("INVALID\n");
                break;
        }
    }
    else
    {
        printf("INVALID\n");
    }
}
        

File: pset6/credit/credit.py
# Import regex
import re
import math

# Get CC num input from the user
# Get user input
while True:
    num = input("Number: ")
    if re.match("^(?=.+)(?:[1-9]\d*|0)?(?:\.\d+)?$", num) is not None:
        break

# Finds out length of Card num
len_num = len(num)

# Define variable for calculating
num = int(num)
count = math.floor(len_num / 2)
num_scale = 10
val = 0

# Calculate from second-to-last digit
for i in range(count):
    sec_digit = (math.floor(num / num_scale)) % 10
    each_val = sec_digit * 2
    if (each_val >= 10):
        each_val = (each_val % 10) + 1
    val += each_val
    if (i == count - 1):
        digit0 = sec_digit
    num_scale *= 100

# Redefine few variable for another calculating
num_scale = 1
count = len_num - math.floor(len_num / 2)

# Calculate with the val of the other digits
for j in range(count):
    digit = (math.floor(num / num_scale)) % 10
    val += digit
    if (j == count - 1):
        digit1 = digit
    num_scale *= 100

# Define variable for validating
check = val % 10

# Validate last digit in the val
if check == 0:
    if len_num == 15:
        if (digit1 == 3):
            if (digit0 == 4 or digit0 == 7):
                print("AMEX")
            else:
                print("INVALID")
        else:
            print("INVALID")

    elif len_num == 13:
        if (digit1 == 4):
            print("VISA")
        else:
            print("INVALID")

    elif len_num == 16:
        if (digit0 == 4):
            print("VISA")
        elif (digit0 == 5):
            if (digit1 >= 1 and digit1 <= 5):
                print("MASTERCARD")
            else:
                print("INVALID")
        else:
            print("INVALID")
    else:
        print("INVALID")
else:
    print("INVALID")